phidgets/manager.go

```diff
@@ -0,0 +1,148 @@
+func attach_handler(man C.PhidgetManagerHandle, ctx unsafe.Pointer, channel C.PhidgetHandle) {
+	m := (*PhidgetManager)(ctx)
+
+	var class C.Phidget_ChannelClass
+	if cerr := C.Phidget_getChannelClass(channel, &class); cerr != C.EPHIDGET_OK {
+		fmt.Printf("unable to determine class of attached phidget: %d\n", cerr)
+		return
+	}
+
+	m.Lock()
+	defer m.Unlock()
+
+	// TODO: All supported phidgets should be here
+	switch class {
+	case C.PHIDCHCLASS_ACCELEROMETER:
+		p := &PhidgetAccelerometer{}
+		p.phidget.handle = channel
+		p.handle = (C.PhidgetAccelerometerHandle)(unsafe.Pointer(channel))
+		m.handles = append(m.handles, p)
+	case C.PHIDCHCLASS_CURRENTINPUT:
+		p := &PhidgetCurrentInput{}
+		p.phidget.handle = channel
+		p.handle = (C.PhidgetCurrentInputHandle)(unsafe.Pointer(channel))
+		m.handles = append(m.handles, p)
+	case C.PHIDCHCLASS_LCD:
+		p := &PhidgetLCD{}
+		p.phidget.handle = channel
+		p.handle = (C.PhidgetLCDHandle)(unsafe.Pointer(channel))
+		m.handles = append(m.handles, p)
+	case C.PHIDCHCLASS_LIGHTSENSOR:
+		p := &PhidgetLightSensor{}
+		p.phidget.handle = channel
+		p.handle = (C.PhidgetLightSensorHandle)(unsafe.Pointer(channel))
+		m.handles = append(m.handles, p)
+	case C.PHIDCHCLASS_SOUNDSENSOR:
+		p := &PhidgetSoundSensor{}
+		p.phidget.handle = channel
+		p.handle = (C.PhidgetSoundSensorHandle)(unsafe.Pointer(channel))
+		m.handles = append(m.handles, p)
+	case C.PHIDCHCLASS_TEMPERATURESENSOR:
+		p := &PhidgetTemperatureSensor{}
+		p.phidget.handle = channel
+		p.handle = (C.PhidgetTemperatureSensorHandle)(unsafe.Pointer(channel))
+		m.handles = append(m.handles, p)
+	case C.PHIDCHCLASS_VOLTAGEINPUT:
+		p := &PhidgetVoltageInput{}
+		p.phidget.handle = channel
+		p.handle = (C.PhidgetVoltageInputHandle)(unsafe.Pointer(channel))
+		m.handles = append(m.handles, p)
+	case C.PHIDCHCLASS_VOLTAGERATIOINPUT:
+		p := &PhidgetVoltageRatioInput{}
+		p.phidget.handle = channel
+		p.handle = (C.PhidgetVoltageRatioInputHandle)(unsafe.Pointer(channel))
+		m.handles = append(m.handles, p)
+	default:
+		fmt.Printf("unsupported phidget discovered: 0x%x\n", class)
+		return
+	}
+
+	// TODO: We are not doing a Phidget_release at any point to get rid of these
+	C.Phidget_retain(channel)
+}
+
+//NewPhidgetManager Create creates a phidget manager
+func NewPhidgetManager() (*PhidgetManager, error) {
+	m := &PhidgetManager{}
+	C.PhidgetManager_create(&m.handle)
+
+	cerr := C.PhidgetManager_setOnAttachHandler(m.handle, (C.attach_fcn)(unsafe.Pointer(C.cattach_callback)), unsafe.Pointer(m))
+	if cerr != C.EPHIDGET_OK {
+		C.PhidgetManager_delete(&m.handle)
+		return nil, managerError(cerr)
+	}
+	// TODO: Should install a PhidgetManager_OnDetachCallback as well so we can
+	// support removing the devices
+
+	if cerr := C.PhidgetManager_open(m.handle); cerr != C.EPHIDGET_OK {
+		C.PhidgetManager_delete(&m.handle)
+		return nil, managerError(cerr)
+	}
+
+	// Sometimes the phidgets take a while to attach
+	time.Sleep(20 * time.Millisecond)
```

Comment:
Maybe instead, this sleep should be left for the user to implement. Going off the example here: https://www.phidgets.com/docs/Phidget_Manager they'd have to sleep if they want to poll with `ListPhidgets` immediately after `open`.

Not having the sleep shouldn't affect the attach handler, right?

Reply:
Yes, that wouldn't be unreasonable. It's annoying that existing Phidgets aren't immediately available, but this is really a feature of the Phidget API, not the Go library, so we should just replicate their functionality. 20ms sleep isn't accurate anyway. I'll remove this and put an explicity sleep in main.go instead, with a comment about it.